### PR TITLE
Clang Format

### PR DESCRIPTION
**Pull request by sslattery**
Adds a clang format file with hooks to run this automatically.

### CODE REVIEW

**Comment by sslattery** [comment]
@junghans Clang format file has been added to this PR

**Comment by sslattery** [comment]
I have reduced the clang format file to only make a few changes to the default LLVM format

**Comment by junghans** [comment]
Ok, I added a `make format` rule and added it to CI, this should fail now.

File: .travis.yml [hunk 1]
```diff
@@ -97,6 +97,7 @@ script:
           ${COVERAGE:+-DCabana_ENABLE_COVERAGE_BUILD=ON -DCOMPILER_SUPPORTS_MARCH=OFF} .. &&
     make -k -j2 VERBOSE=1 &&
     make test CTEST_OUTPUT_ON_FAILURE=1 &&
+    make format && git diff --exit-code &&
```
Comment (dalg24) [inline .travis.yml]: Nice

File: CMakeLists.txt [hunk 2]
```diff
@@ -133,3 +133,11 @@ add_subdirectory(core)
 
 configure_file(cmake/Cabana.pc.cmakein Cabana.pc @ONLY)
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/Cabana.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
+
+find_package(CLANG_FORMAT 7.0.1)
+if(CLANG_FORMAT_FOUND)
+  file(GLOB_RECURSE FORMAT_SOURCES core/*.cpp core/*.hpp)
```
Comment (dalg24) [inline CMakeLists.txt]: I would have gone for something like `git ls-files | grep -E "\.hpp$|\.cpp$"` but that works.

Comment (junghans) [reply CMakeLists.txt]: Yeah, I just wanted to avoid the usage of more external tools like git and grep.

**Comment by junghans** [comment]
> Find module for clang-format feels a bit overkill but this is good overall.

It is, but at least now we can require a minimum version if needed.

**Comment by junghans** [comment]
@sslattery if the formatting is fine with you, let’s merge this.

**Comment by junghans** [comment]
ping @sslattery

File: core/src/Cabana_Parallel.hpp [hunk 3]
```diff
@@ -20,33 +20,27 @@
 #include <cstdlib>
 #include <type_traits>
 
-namespace Cabana
-{
+namespace Cabana {
```
Comment (sslattery) [inline core/src/Cabana_Parallel.hpp]: This is probably the main thing I don't like - having an open `{` or `(` not start on the next line.

File: .clang-format [hunk 4]
```diff
@@ -0,0 +1,8 @@
+BasedOnStyle:  LLVM
+---
+Language:        Cpp
+AlwaysBreakTemplateDeclarations: true
+BreakBeforeBraces: Custom
```
Comment (sslattery) [inline .clang-format]: Would this get changed to true to allow for the break I mentioned below?

Comment (junghans) [reply .clang-format]: `true` don't work, but `Allman` does.

File: core/src/Cabana_Parallel.hpp (hunk 3 above)
Comment (junghans) [reply core/src/Cabana_Parallel.hpp]: Fixed!

**Comment by junghans** [comment]
The remaining differences seem to be different defaults in `clang-format-7` and `clang-format-8`. What style do you prefer?

**Comment by sslattery** [comment]
Nice! For me I have no preference of 7 vs. 8. If the difference is not substantial I would say go with the latest (8).

**Pull request merged**